refactoring the plotting system

wpfff opened the pull request:
some first working bits and pieces for slightly restructured plotting code.
main aspects:
- separate all parts that are not backend (like matplotlib) dependent but more general. This includes quite a lot of the current logic in the autoplot elements.
- should end up such that adding other plotting backends should be fairly easy.

Some notes for orientation:
- there's no more mpl.py file but rather a submodule
- included a simple way for customizing matplotlib appearance using a config system with ``plottrconfig_*.py`` file
- added a new class 'AutoFigureMaker' whose job is only to track the elements of the figure and arrange them into subplots depending on user settings.
- 2d plots are not yet working, but it should be enough to get an idea of where it's going.

wpfff committed:
started working on better plot system.

wpfff committed:
merge from main repo.

wpfff committed:
added some basic functionality to find config files.

wpfff committed:
small fix: allow passing of keyword arguments from the command line.

wpfff committed:
[WIP] started to move matplotlib helper functions around.

wpfff committed:
fix: don't import from lower level module on import to avoid circular imports.

wpfff committed:
added stylefile where we can easily find it.

wpfff committed:
moved things up a level that are basically plotting backend-independent.
Added new class that is supposed to handle arranging plots into different panels (no need to mix that with actual plotting code).

wpfff committed:
moved widgets that use MPL but are higher-level than autoplot into here.

wpfff committed:
[WIP] first steps and in modified autoplot using matplotlib. so far only 1D plots work, rest is non-functional at this point.

wpfff committed:
merge from remote master.

wpfff commented:
@jenshnielsen 
this is still very rough, but i thought i'll open a PR already to give you chance to look since you also wanted to make changes to the matplotlib parts.

some tests are working fine, which hopefully should be enough to get an idea, see:
/test/gui/mpl_autoplot.py
/test/apps/autoplot_app.py

(2D plots i need to plug back in -- i stopped today when the main parts seemed to fit together)

would be great to hear any thoughts!

wpfff commented:
i should also add that the recent changes for high-dpi scaling aren't yet re-implemented here. but shouldn't be too hard to get that merged back in.

jenshnielsen commented on plottr/plot/base.py:
```diff
@@ -134,3 +144,315 @@ def makeFlowchartWithPlot(nodes: List[Tuple[str, Type[Node]]],
+    #: unplottable data
+    unknown = auto()
+
+    #: scatter-type data with 1 dependent (data is not on a grid)
```
Strictly I guess this should be that data is not sorted. It's fine to do a 1d line plot for data with variable step size

wpfff replied:
yes, agreed.

jenshnielsen commented on plottr/plot/base.py:
```diff
@@ -134,3 +144,315 @@ def makeFlowchartWithPlot(nodes: List[Tuple[str, Type[Node]]],
+    axes: Optional[List[Any]] = None
+
+
+class AutoFigureMaker(object):
```
Inheriting from object is redundant and does nothing in python 3

wpfff replied:
old habits... :)

astafan8 commented on plottr/plot/base.py:
```diff
@@ -134,3 +144,315 @@ def makeFlowchartWithPlot(nodes: List[Tuple[str, Type[Node]]],
+            if len(plotItem.data) == 2:
+                style = 'line'
+            elif len(plotItem.data) == 3:
+                style = 'image'
```
what's the difference between this "style" and "PlotType" enum? also, shouldn't determinePlotDataType be used here somehwere?

jenshnielsen replied:
As I understand it style is basically:

```1D or 2D```

```PlotDataType``` and ```PlotType``` are further refinements of this. 

Perhaps we should consider if we can streamline these types. My understanding is basically that we can categorize the data as follows. I like to think of it as a mapping from data types to valid plot types

* 1d data ordered: [1D line plot, 1D points plot (scatter)]
* 1d data not ordered: [1D points plot (scatter)]
* 2d data on equidistant grid: [image, colormesh, scatter2d]
* 2d data on non equidistant grid: [colormesh, scatter2d]
* 2d data no structure: [scatter2d]

Further to that there is choice between single and multitrace 1d plots which seems to be somewhat unrelated.

it seems to me that these types currently mix things that should be separate concerns but since that code is the same as it was before in mpl.py we should probably not touch that in this pr

wpfff replied:
good point, Mikhail. this could essentially be the same thing. the idea is that any backend could essentially support different types of plots that for instance also the user may want to implement (i'm thinking of a few custom plots that would be useful for specific measurements).  Other backends might support 3D graphics in a decent way. etc...

Jens, agreed. The only thing i'd want to make sure of is that adding plot types should be somewhat easy, and potentially doable through a plugin rather than only by modifying 'core' code. 

I'm definitely open to come up with a structure that does things 'right' in this effort, in terms of separation. There's no immediate rush with merging this into the main branch.

astafan8 commented on plottr/plot/base.py:
```diff
@@ -134,3 +144,315 @@ def makeFlowchartWithPlot(nodes: List[Tuple[str, Type[Node]]],
+    def plot_line(self, plotItem: PlotItem):
+        raise NotImplementedError
+
+    def plot_image(self, plotItem: PlotItem):
```
shouldn't there be methods for each PlotType ?

wpfff replied:
Good idea, this would probably be cleaner. See also comment above: i'd need to think a bit on how to make sure this remains extendable, and there still needs to be a way to translate this to multiple backends.

astafan8 commented on plottr/plot/mpl/__init__.py:
```diff
@@ -0,0 +1,44 @@
+    """Set the matplotlib style based on available config files."""
+    styleFiles = configFiles('plottr_default.mplstyle')
+    if len(styleFiles) > 0:
+        plt.style.use(styleFiles[0])
```
if there are more the one files, then perhaps they should be applied in a "hierarchical sum" like, say, git config files?

wpfff replied:
yes, that'd be great. i have to mess around with rc params in matplotlib a bit to see how that could work.

wpfff replied:
So far I don't see an easy way with the matplot methods directly.
I'm thinking of including this into a central config file for plottr that will also configure matplotlib parameters.

wpfff replied:
i pushed a proposal for a config system with just python files (so no other specific file format) that both supports the hierarchical sum concept you mentioned, and can do all configuration of matplotlib.

astafan8 commented on plottr/plot/mpl/autoplot.py:
```diff
@@ -0,0 +1,505 @@
+            PlotType.singletraces: self.plotasSingleTraces,
+            PlotType.image: self.plotasImage,
+            PlotType.colormesh: self.plotasMesh,
+            PlotType.scatter2d: self.plotasScatter2d,
```
exactly, shouldn't these functions be required to be implemented by a concrete AutoFigureMaker class?

astafan8 commented on plottr/plot/mpl/autoplot.py:
```diff
@@ -0,0 +1,505 @@
+        """Plot the data using previously determined data and plot types."""
+
+        if self.plotDataType is PlotDataType.unknown:
+            logger.debug("No plotable data.")
```
missing early return?
```suggestion
            logger.debug("No plotable data.")
            return
```

jenshnielsen reviewed (commented):
@wpfff I think this looks promising. There is a bit of work to ensure feature parity with the existing code and some cleanup but I like the general direction

wpfff commented:
Thanks for the feedback, both of you! I'll probably have some time this weekend to ponder some of the deeper suggestions and see if i can refine the concept a bit (especially the role of the PlotType, style, etc.).

wpfff committed:
Merge branch 'master' into betterplotting

wpfff committed:
Merge branch 'master' into betterplotting

# Conflicts:
#	plottr/__init__.py
#	plottr/plot/mpl.py

wpfff committed:
some better organization:
- PlotType is not a global concept, but something only the MPL plot uses
- PlotItem.style is replaced with type PlotDataType to streamline
- make no assumptions about plotting, leave all details to the backend classes

wpfff committed:
clarifying better which objects are responsible for what:
- PlotWidgets obtain data from the Nodes, and process UI options
- based on UI options, data is processed and passed to FigureMaker
- No actual plotting happens in the Widget, this is the role of the FigureMaker

wpfff committed:
re-enabled basic colorplotting.
deleted old templates.

wpfff committed:
fixed type hints in plotting.py

wpfff committed:
fixing more typing annotations for mypy.

wpfff committed:
more cleanup for mypy.

wpfff committed:
more cleanup for mypy.

wpfff committed:
more cleanup for mypy.

wpfff committed:
more cleanup for mypy.

wpfff committed:
emit windowClosed signal for all PlotWindow objects.

wpfff committed:
improved testing of the autoplot app.
easier to add new tests, supports testing live plotting.

wpfff committed:
fixed tests to work with latest changes on FigureMaker.

wpfff committed:
- make sure we plot in the correct panels when plotting complex data in different representations.
- minor fixes to avoid warnings

wpfff committed:
minor fixes and aesthetic improvements

wpfff committed:
Updated docs for the base plotting module.

wpfff committed:
minor fixes (broken links).

wpfff committed:
completed docs for plotting modules.

wpfff committed:
added font scaling for high-dpi.

wpfff commented:
@jenshnielsen @astafan8  -- with a slight delay caused by a winter full of grant writing :| 

I think i addressed most of the stuff you had brought up before, i think it's a bit cleaner now.
(and fully functional, as far as I've tested).

FarBo commented:
@wpfff 
It seems this refactoring lacks the option we implemented a while back to have complex options (Real, Real/Imag, Mag/Phase) in the plot window.
So, it only has a button callded Mag/Phase, which by default is disabled (default is plotting Real/Imag) and enabling plots Mag/Phase. 
I tried to implement these options, but I faced a few problems. I am creating a PR in this branch with my modifications and expalin what the problems are.

FarBo committed:
Full complex options buttons

wpfff commented:
@FarBo do you mean it's missing an option to just plot Real? Because the choice between Real/Imag and Mag/Phase is implemented. (I tried to streamline the number of buttons a bit, like using toggle buttons when it comes to effectively binary choices)

wpfff commented:
@FarBo you could checkout the file tests/apps/autoplot_app.py for some examples i played with.

FarBo commented:
@wpfff 

> do you mean it's missing

Yes, if we want to have real plot, then we need to have 3 buttons, if not, the current implementation works.
Also, in my implementation before this refactoring, the real/imag plot was plotting real and imag parts separately in two subplots, which I think it is more cleaner.
I am trying to push my changes in a PR in your forked repo, then we can discuss there easily, but it rejects me to create a PR there and says Permission is denied :)

wpfff commented:
@FarBo I see - plotting real and imaginary separately is honestly something i hadn't thought about -- for none of my own needs i would see this as desirable, but i see how that could be useful to others :)

to summarize: we should then maybe have an option to only plot real, and an option to separate real/imag into different subpanels?

wpfff commented:
hm, i can't figure out how to give you access to my fork... i'll try to figure that out today.

FarBo commented:
@wpfff 
I agree with the summary.
I did most of things regarding having real, so I can create a PR, then we can discuss there.
I don't know on top of my head to how to give access. But, please let me know if you made it possible :)

wpfff commented:
i can't figure out how to do this generally.... i invited you as a collaborator to my fork, maybe that will work?

FarBo commented:
it worked. i created the pr there

wpfff committed:
- fix plotting of Real part only
- added option for separating Re/Im

wpfff commented:
@FarBo i adopted your suggestions, it all seems to work fine.
I'll leave making the UI prettier (with icons for each options) for a later PR.

FarBo commented:
@wpfff 
Thanks for further work on that. That's correct and seems it is working :). There is an extremely tiny thing that I just want to report and it is not like a bug or issue in my opinion: For colormap plots, Re/Im and split Re/Im are the same and should be the same of course (In a very ideal world, one of them could be disabled. But the amount of work to have this little thing does not worth it)

Regardless of this complex topic, I noticed a few things that I would like to mention here (I highlighted the differences between before and after refactoring that might worth to notice and improve):
Pic1(copy metadata is missing, plot title is on top of plot and plot icons look very small):
https://microsoft-my.sharepoint.com/:i:/p/a-fbonabi/Ef0aHJvCXu1HiIu1GwHgiZ4BfgmcxECSL8aTe_GDvn-CMA?e=xlUseN
Pic2 (the legend is in the middle of plot for this specific case):
https://microsoft-my.sharepoint.com/:i:/p/a-fbonabi/EeAIZkILEpNDsjrWWkHdg_8BHhPF-Dh-8yC8xjKxcG4DJg?e=GA5TNp

FarBo commented:
@wpfff
Regarding the mentioned points above:
I think Metadata export added in this PR by @astafan8:
https://github.com/toolsforexperiments/plottr/pull/173

Scaling is also handled in this PR by @Akshita07:
https://github.com/toolsforexperiments/plottr/pull/156

wpfff commented:
@FarBo i don't have access to the screenshots. can you embed somewhere else than sharepoint?

FarBo commented:
@wpfff 
Sorry, I chose my work onedrive to share :)
This should work:
https://1drv.ms/u/s!AkKmOu47nt6ygrI-hkeJZ-1ZzjU6Kw?e=6vX4Ob

wpfff commented:
ah, yes -- i think i missed a couple of things that were implemented while i worked on this. i'll try to fix that today.

wpfff committed:
reimplemented dpi scaling in all widgets and copying meta info.

wpfff commented:
I've reimplemented those two pieces.
One comment about copying the meta info: I really like that feature, but think it'd be better if it copied all metadata rather than just the qcodes specific fields (that's what i've done now).

For instance, I'll be using the tool without the qcodes dataset mostly, so my meta data will have different keys (but i would still like to have a button that supports that).
I think it makes more sense to copy everything regardless of where the meta data comes from.

If you really only want to copy the qcodes entries, we could think of a way to filter (maybe through a config file or a variable that gets set when then autoplot app is created).

astafan8 commented:
> One comment about copying the meta info: I really like that feature, but think it'd be better if it copied all metadata rather than just the qcodes specific fields (that's what i've done now).

this is a great suggestion! let's indeed copy all metadata. however,
1. the thing in the original implementation was to strip the qcodes_ prefix for the keys, otherwise the resulting keys in the metadata will have this annoying (?) prefix. one way would be to remove that prefix from the place where these keys are created (upon reading of qcodes dataset), another way would be to keep the logic that strips the prefix from the keys. or, we say it's a non-issue.
2. another thing is - are the values of the "meta info" guaranteed to just be small strings and numbers? or can they also be numpy arrays or something huge? if so, then filtering the keys to copy is actually useful.

wpfff commented:
>     1. the thing in the original implementation was to strip the qcodes_ prefix for the keys, otherwise the resulting keys in the metadata will have this annoying (?) prefix. one way would be to remove that prefix from the place where these keys are created (upon reading of qcodes dataset), another way would be to keep the logic that strips the prefix from the keys. or, we say it's a non-issue.

Good point, i thought about that, but i think it boils down to a matter of taste. My thinking is that there can be multiple 'sources' of metadata (a student i'm working with is working on a fitting module, for example), and the prefix would actually help group entries by meaning. I'm tempted to vote non-issue for now.
 
>     2. another thing is - are the values of the "meta info" guaranteed to just be small strings and numbers? or can they also be numpy arrays or something huge? if so, then filtering the keys to copy is actually useful.

Good catch. I could imagine to either only allow 'simple' data types, or to return the string representation and truncate after N characters (maybe 200 or so).

wpfff committed:
make sure meta entries don't get too long when copying.

wpfff committed:
don't inherit from object.

wpfff committed:
- added central method for scaling dpi
- switch to configuration method that allows hierarchical resolution of user configuration (pure python, because that seems the best solution for configuring matplotlib)

FarBo commented:
@wpfff 
I see copy metadata is added. Legend seems behaving fine.
Still the plottype buttons are very small compared to the latest release:
https://1drv.ms/u/s!AkKmOu47nt6ygrI_W8T8SJUw_AHZFQ?e=ArafkO

wpfff commented:
> Still the plottype buttons are very small compared to the latest release:
> https://1drv.ms/u/s!AkKmOu47nt6ygrI_W8T8SJUw_AHZFQ?e=ArafkO

@FarBo : true -- matter of preference I guess. Along with one of my recent commits I made them the same size as the ones in the other toolbar to be consistent which i thought looked better (they are dpi-scaled, though).
I'm open to changing that back, though, if you disagree.

FarBo commented:
@wpfff
`true -- matter of preference I guess. Along with one of my recent commits I made them the same size as the ones in the other toolbar to be consistent which i thought looked better (they are dpi-scaled, though).
I'm open to changing that back, though, if you disagree.`

The buttons on the right looks natural to me and as far as I remember, small sizes (left) are not preferred by some of users.
I think it would be nice to have the bigger like the right plot.

wpfff committed:
change back to bigger icons in the toolbar

wpfff committed:
fixed a couple of mypy issues.

wpfff committed:
more mypy fixing.

FarBo commented:
@wpfff 
Thanks for the updates. I didn't go through the code structure and only tested the functionality. These final commits solved all my raised concerns. :)

wpfff commented:
> @wpfff
> Thanks for the updates. I didn't go through the code structure and only tested the functionality. These final commits solved all my raised concerns. :)

Thanks, @FarBo! I'm still waiting for comments from @astafan8 (for this PR i think it'd be important that there's some consensus on structure, and he wanted to have a look)

wpfff commented:
@astafan8 - just to poke you :) are you still planning on giving comments on this?
If not, i'd like to merge soon. It's quite well-behaved now and i'm working on an additional backend that follows this structure.
There will probably still be some refactoring once i have another backend going (mainly regarding what should be part of the matplotlib specific part and what is more general) but the overall structure works well i think.

astafan8 commented:
> just to poke you :) are you still planning on giving comments on this?

thanks @wpfff ! I will do this week

astafan8 commented:
@wpfff ETA today :)

astafan8 committed:
Add disappeared "imported plottr version" log message

astafan8 commented on plottr/__init__.py:
```diff
@@ -15,11 +17,97 @@
+            if forceReload:
+                importlib.reload(mod)
+            this_cfg.update(getattr(mod, 'config', {}))
+            sys.path.pop(0)
```
[just dropping a note, not suggesting to change it in this PR] not a fan of touching the path (even temporarily), so i looked up a bit and found this recipe in python documentation using importlib https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly (also on stackoverflow https://stackoverflow.com/questions/67631/how-to-import-a-module-given-the-full-path) - i think this would've been cleaner/safer. Another example is how sphinx does it with their config.py files https://github.com/sphinx-doc/sphinx/blob/e878299cc608614e959e0e99025cccfd48de00bc/sphinx/config.py#L316 using compile/exec resulting in a dictionary, which might be a bit overkill given that importlib can do everything.

wpfff replied:
i like this better than the current solution, and is very simple. i changed that.

astafan8 commented on plottr/plot/base.py:
```diff
@@ -120,17 +129,482 @@ def __init__(self, parent: Optional[QtWidgets.QWidget] = None) -> None:
+
+        :param subPlotId: ID of the subplot.
+        :return: Depends on inheriting class.
+        """
```
missed?
```suggestion
        """
        raise NotImplementedError
```

wpfff replied:
i think this one does not really need to be implemented, and a subclass could choose not to.
In that case we would not want it to raise an exception. 
i clarified in the docstring.

astafan8 commented on plottr/plot/mpl/widgets.py:
```diff
@@ -0,0 +1,209 @@
+
+        self.addMplBarOptions()
+        scaling = rint(self.logicalDpiX() / 96.0)
+        defaultIconSize = 16 * scaling
```
```suggestion
        defaultIconSize = 16 * dpiScalingFactor(self)
```

astafan8 commented on plottr/plot/mpl/autoplot.py:
```diff
@@ -0,0 +1,466 @@
+        assert len(plotItem.data) == 2
+        lbl = plotItem.labels[-1] if isinstance(plotItem.labels, list) and len(plotItem.labels) > 0 else ''
+        x, y = plotItem.data
+        return axes[0].plot(x, y.real, label=lbl, **plotItem.plotOptions)
```
what is `.real` really needed for? isn't the data already split into real/imag/mag/phase parts? same in the plotImage method
```suggestion
        return axes[0].plot(x, y, label=lbl, **plotItem.plotOptions)
```

wpfff replied:
good point -- this looks wrong (although there was still a -- hidden -- possibility that data might be complex typed).
i fixed it in the correct spot (apparently numpy.ma.abs can produce complex valued data).

astafan8 reviewed (approved):
left a few notes/comments but let's merge this finally :) i'm happy to tag a release tomorrow

astafan8 committed:
Merge branch 'master' into betterplotting

wpfff commented:
Thanks, @astafan8 -- i'll adopt your suggestions today. 
I'll probably also include the importlib suggestion. Much better solution -- i just hadn't looked properly :)

wpfff committed:
importing config files without modifying sys.path.

wpfff committed:
clarify why `formatSubPlot` does not raise an error when not reimplemented.

wpfff committed:
fix mypy issue with importlib.

wpfff committed:
Update plottr/plot/mpl/widgets.py

Co-authored-by: Mikhail Astafev <astafan8@gmail.com>

wpfff committed:
enforcing real in a more sane spot.

wpfff committed:
Merge remote-tracking branch 'origin/betterplotting' into betterplotting

wpfff committed:
fix typing.
